### ISSUE TITLE
MD5 changes by operating system

### ISSUE THREAD

**Issue by veli-m-andirin** [gslab-econ/gslab_r#15]
In another project we noticed that MD5 hashes change for the same R object depending on the operating system. We also noticed the same issue for outputted CSV files.

When computing hashes on outputted CSV files, we were able to get a unique hash by specifying the same line ending for different operating systems.

The issue remains for hashes computed inside R.

@santiagohermo, @miikapaal fyi.

**Comment by veli-m-andirin** [comment]
@santiagohermo @miikapaal,

In term of obtaining a unique hash on different operating systems using SaveData, I suggest that we save CSVs using the same line ending across platforms, and compute hashes on the outputted files.
 
Would you agree?

**Comment by miikapaal** [comment]
> @santiagohermo @miikapaal,
> 
> In term of obtaining a unique hash on different operating systems using SaveData, I suggest that we save CSVs using the same line ending across platforms, and compute hashes on the outputted files.
> 
> Would you agree?

Thanks @veli-m-andirin! Yes, that sounds like a reasonable starting point.  I've just created a local version of the `gslab_r` repo.

**Comment by santiagohermo** [comment]
I agree with [this idea](https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-921788493) @veli-m-andirin! My only comment would be that I would make that a non-default option. If you are working in a single machine with a large dataset (that'll take a while to save), then you are probably ok with just using plain digest.

I think it'd also be cool to figure out why R objects (before saving) produce different hashes, but given our previous exploration I wouldn't know where to start.

**Comment by veli-m-andirin** [comment]
@miikapaal @santiagohermo reviving this thread! (Please feel free to look at this when you have bandwidth, potentially after Skills production ends)

@santiagohermo, to compute the MD5 hash for a file, we might actually not need to load it into R's memory, at least in Windows. [This function](https://www.rdocumentation.org/packages/tools/versions/3.6.2/topics/md5sum) did the computation rapidly in my case.

I am worried that this might not be the case for Mac or Linux. @miikapaal, when you get a chance, could you compute the MD5 hash of a large file using [this function](https://www.rdocumentation.org/packages/tools/versions/3.6.2/topics/md5sum), and see if it is faster than the time it would take to load this file into R?

**Comment by veli-m-andirin** [comment]
Hi @miikapaal @santiagohermo! https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1013146841 might have been forgotten, so I wanted to ping you :smile:

**Comment by santiagohermo** [comment]
A little program might help here @veli-m-andirin @miikapaal, so that it's clearer what we are comparing

When I run the below in RStudio I get

```R

N = 1000000
df <- data.table(id = seq(1, N),
                 x  = rnorm(N))


system.time(write.csv(df, file = "df.csv", row.names = F))

library(data.table)
setDT(df)

system.time(fwrite(df, file = "df.csv"))

system.time(tools::md5sum("df.csv"))

```

I get

![image](https://user-images.githubusercontent.com/45404755/151677039-149c8dda-6005-42f2-a1c7-df56ac7dfa96.png)

**Comment by veli-m-andirin** [comment]
Thanks @santiagohermo!

@miikapaal, is it possible for you to post what you get if you run the script in https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1024983543 as well?

**Comment by miikapaal** [comment]
@veli-m-andirin @santiagohermo Sorry for being slow here!

The code in https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1024983543 produced an error (since `data.table` wasn't loaded when creating `df`), so I moved `library(data.table)` to appear first. In other words, I ran the following in RStudio (R version 4.0.5):
```R
library(data.table)

N = 1000000
df <- data.table(id = seq(1, N),
                 x  = rnorm(N))


system.time(write.csv(df, file = "df.csv", row.names = F))

setDT(df)

system.time(fwrite(df, file = "df.csv"))

system.time(tools::md5sum("df.csv"))
```
which produced

<img width="513" alt="Screenshot 2022-03-13 at 20 59 03" src="https://user-images.githubusercontent.com/65957491/158076808-b4f4e29b-864f-48b5-9296-35de7eae6399.png">

**Comment by veli-m-andirin** [comment]
Thanks @santiagohermo @miikapaal! Since computing the md5sum on an outputted object seems to be quite fast, I think I can implement the proposal in https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-921788493 if you'd agree.

**Comment by santiagohermo** [comment]
Sounds good @veli-m-andirin! Quick question: what would the function do if the user does not specify saving the file as csv?

**Comment by veli-m-andirin** [comment]
Thanks @santiagohermo! I think it should keep doing what it previously did for non-csv files. If we also have issues with changing MD5 hashes for other file types, it might make sense to track those later, since our previous exploration was for CSVs. Would you agree with that?

**Comment by santiagohermo** [comment]
Thanks @veli-m-andirin! Are we sure this problem is particular to csv files? I think it'd make sense to implement a solution to all file types, just to be safe.

**Comment by miikapaal** [comment]
Thanks @santiagohermo @veli-m-andirin!

> Thanks @veli-m-andirin! Are we sure this problem is particular to csv files? I think it'd make sense to implement a solution to all file types, just to be safe.

I agree with @santiagohermo here. I would assume that checking all relevant file types is straightforward?

**Comment by miikapaal** [comment]
As part of another project, we noticed that the MD5 hash might also differ across machines because e.g. Mac and Windows sort rows of a data set slightly differently.

**Comment by santiagohermo** [comment]
To complement [the above](https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1070785611), in this project the different sorting emerged when the column we sorted on was a string, and the first word of the string ended with a hyphen.

**Comment by veli-m-andirin** [comment]
Thanks @santiagohermo @miikapaal!

On https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1066312126:

I agree that it would be good to check other files as well. I uploaded a script in d254f0e. If @miikapaal can run it on his Mac maybe we can get an idea about which file types are likely to have unstable MD5 hashes.

One caveat is that I don't have much idea about binary files. For CSV's, we have more options to control how they are structured, e.g. we can set how the line endings should be. For other file formats, the functions that we use provide less options. Thus, even if we understand the problem, we may not be able to solve it. 

Running this script might be good at least for documentation purposes.

@santiagohermo, can you also run the script? You'll notice that it saves the file in .RData format, but throws an error. When I ran traceback(), what I got wasn't very informative since for some reason the entire data frame is printed after the do.call() step in the code. We may need to open an issue to see if there's a bug with how we use .RData format.

On this https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1070785611:

Do you have any idea about how to proceed with this? The only thing I thought of is to try to impose a custom rule for sorting data frames, which sounds like an overkill (even if we learn how to do it).

**Comment by santiagohermo** [comment]
Thanks @veli-m-andirin !

> @santiagohermo, can you also run the script? You'll notice that it saves the file in .RData format, but throws an error.

I ran the script and got a change in data_manifest. This is not surprising though, since we are simulating data without setting a seed. I don't have access to this repo, but I suggest the following updates to the code:

<details><summary>Code</summary>
<p>

```R

remove(list=ls())
library(SaveData)

set.seed(0)

n <- 10000
df_1 <- data.frame("a" = round(runif(n, 0, 1), 2), 
                   "b" = round(rnorm(n, 5, 2), 2),
                   "c" = sample(c("x", "y", "z"), n, replace = T),
                   id = 1:n)

SaveData(df_1, key = "id", outfile = "issue15/df_1.csv",
         logfile = "issue15/df1_manifest.log")

SaveData(df_1, key = "id", outfile = "issue15/df_1.RDS",
         logfile = "issue15/df1_manifest.log", appendlog = T)

SaveData(df_1, key = "id", outfile = "issue15/df_1.dta",
         logfile = "issue15/df1_manifest.log", appendlog = T)

SaveData(df_1, key = "id", outfile = "issue15/df_1.RData",
         logfile = "issue15/df1_manifest.log", appendlog = T)
```

</p>
</details>

Regarding the error. I also got this! I think I would drop the possibility of saving in `.RData` format altogether (I don't see any moment in which you may want to use it to save a single file as we have `.RDS`)

> On this [#15 (comment)](https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1070785611):
> 
> Do you have any idea about how to proceed with this? The only thing I thought of is to try to impose a custom rule for sorting data frames, which sounds like an overkill (even if we learn how to do it).

We would need to change [the sorting procedure](https://github.com/gslab-econ/gslab_r/blob/d4038209865e7caf2026267b843381d4c0404b4f/SaveData/R/SaveData.R#L101-L110) to use a stable sort function. Maybe we can try `data.table`'s [`setorder`](https://www.rdocumentation.org/packages/data.table/versions/1.14.2/topics/setorder)? The whole function would probably work faster if we use `data.table` in [here](https://github.com/gslab-econ/gslab_r/blob/d4038209865e7caf2026267b843381d4c0404b4f/SaveData/R/SaveData.R#L158) as well.

fyi @miikapaal

**Comment by miikapaal** [comment]
> I agree that it would be good to check other files as well. I uploaded a script in [d254f0e](https://github.com/gslab-econ/gslab_r/commit/d254f0e5b72768572c6057bbacc379db16130982). If @miikapaal can run it on his Mac maybe we can get an idea about which file types are likely to have unstable MD5 hashes.

Thanks @veli-m-andirin @santiagohermo! I ran the script as well (I guess it's now the version in https://github.com/gslab-econ/gslab_r/commit/749ad6b73ac1fcdf61f6d22a1951576f5c2dc5df). Although the data sets are saved, I get the following error in the last step. Not sure if that's expected?
<img width="518" alt="Screenshot 2022-04-10 at 18 04 38" src="https://user-images.githubusercontent.com/65957491/162628597-9d2a65b7-4e10-4ca9-a29a-99967b1fd8bc.png">

@veli-m-andirin Could you rerun the script and commit the manifest file? Since we're now setting a seed, we can compare which filetypes have unstable hashes.

**Comment by veli-m-andirin** [comment]
Thanks @santiagohermo @miikapaal for great inputs! (And sorry for not having set the seed, my bad).

@miikapaal, I posted the data file manifest in 22c5aee. Could you also share what you get? 

@santiagohermo, on the point about the issue of sorting mentioned in https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1094106202, I'll check it soon.

**Comment by miikapaal** [comment]
> @miikapaal, I posted the data file manifest in [22c5aee](https://github.com/gslab-econ/gslab_r/commit/22c5aeeef45ee9d6f617ca86a77b2b6969bbba74). Could you also share what you get?

Great, thanks @veli-m-andirin! Just ran the script. The manifest remains identical but I still get the same error that I mention in https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1094304392. Do you get the same error? (I'm wondering is this is just a machine-specific thing.)

**Comment by veli-m-andirin** [comment]
@miikapaal, 

I think that's the error about RData that I mentioned in https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1087076122. I agree with @santiagohermo's suggestion drop it all together. 

@santiagohermo @miikapaal could you point me to files where there were changing MD5 hashes in Skills? If they are not too large, maybe we can store them in the issue sub, read them and output them in different formats, just to understand if we have these problems only with CSVs.

**Comment by veli-m-andirin** [comment]
@miikapaal, could you run the script `check_data_types.R` again?

**Comment by miikapaal** [comment]
> @miikapaal, could you run the script `check_data_types.R` again?

@veli-m-andirin Just reran the script, this time without errors. Surprising results: the `manifest` files (i.e. also MD5 hashes) remain identical, even though according to `git` the following files are modified:

<img width="514" alt="Screenshot 2022-04-11 at 11 07 00" src="https://user-images.githubusercontent.com/65957491/162703511-234b9bfb-8282-453e-aa86-c5bf344075cd.png">

Since all of the above are binary files, I don't think I can easily determine how they have been modified. I can of course have a closer look if needed.

**Comment by veli-m-andirin** [comment]
Thanks a lot @miikapaal! When you get a chance, can you push those results? I wonder if GitHub would also claim any differences.

**Comment by veli-m-andirin** [comment]
Thanks @miikapaal! @jmshapir, could you please enable write access for @miikapaal to this repo?

**Comment by miikapaal** [comment]
@veli-m-andirin I've now added the changes from my run, see https://github.com/gslab-econ/gslab_r/commit/778a1139658bf71d25c11a557c0184fc0f9405b5.

**Comment by veli-m-andirin** [comment]
Thanks @miikapaal!

@santiagohermo, based on what @miikapaal posted, it looks like there are some subtle changes in files that can not be detected by even MD5 hashes but can be detected by Git. Maybe the difference stems from the underlying sorting. I think I'll try to implement the switch to data.table's stable sort function and we can run this experiment again.

@santiagohermo @miikapaal please let me know what you think.

**Comment by santiagohermo** [comment]
Thanks @veli-m-andirin! I think we should understand better how GitHub handles binaries. If I remember correctly, when working with pdfs git detects a change after new compilations even if the pdf is identical.

**Comment by veli-m-andirin** [comment]
Thanks @santiagohermo! I think it might also be a good idea to initialize git lfs in the repo. I think it handles binaries differently, and since we are using it for binary files it would be nice to see if the same problem occurs using lfs. 

@miikapaal, would it be possible for you to update your local copy of the repo, run `check_data_types.R`, and see if the files change?

I also think if three of us could have a short meeting to think about how to proceed with this issue and what our goals should be. If you would be fine with it, I will send an email to coordinate.

**Comment by miikapaal** [comment]
> @miikapaal, would it be possible for you to update your local copy of the repo, run `check_data_types.R`, and see if the files change?
> 

I reran the script in https://github.com/gslab-econ/gslab_r/commit/4f73ea50614a019b7e5c91e3420057ec662aecbb. The outcome is the same as last time, i.e. git indicates the binary files were modified even though the manifest logs remain identical. I think the problem is related to https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1097367359.

> I also think if three of us could have a short meeting to think about how to proceed with this issue and what our goals should be. If you would be fine with it, I will send an email to coordinate.

That would work for me. Feel free to send an email to get the ball rolling.

**Comment by jmshapir** [comment]
@veli-m-andirin can you summarize our status here for me? Thanks!

**Comment by veli-m-andirin** [comment]
Thanks @jmshapir!

We started this issue to investigate MD5 changes in the data file manifests of outputted CSV files. Our first thought was that this was caused by different line endings used by operating systems. Then, another observation came from Skills, where it was suspected that some MD5 changes were caused by different methods of sorting rows across used by operating systems.

In addition to these, we started to discuss why Git shows some binaries as changed despite the fact that they have not. We noticed that this issue is relevant both for git and git-lfs. We haven't found any reasons to these yet.

Currently I see two ways of proceeding efficiently here:

1) Implement the two "potential fixes" for CSV files: Adopt a common line ending for all operating systems, and also use a stable sorting function. These might reduce the number of times we see a change in MD5 hashes.
2) Leave everything as is.

Both options suggest leaving the changes with binary files as is. Both git and git lfs find some changes in binaries when they are created on different operating systems. I couldn't find a proper manual on this issue after spending approximately an hour (hence I haven't found anything on the solution either)., I just found people complaining about it. 

My general understanding is that, as long as the variable summaries in data file manifests do not change (means, standard deviations etc.) then we will be fine unless we are looking at character variables. At the same time, we have some seemingly "harmless" changes to SaveData such as using a common line ending and a stable sorting function which we can implement. (I see the latter as more "harmless" as I am not sure if there would be any issues with imposing the same line ending on all operating systems). 

Please let us know your thoughts.

@miikapaal @santiagohermo fyi.

**Comment by jmshapir** [comment]
Thanks @veli-m-andirin!

* On imposing a stable sort rule, this sounds like a good idea.
* On imposing a common line ending, have we seen examples where we get a different MD5 hash on different machines when we do not impose a common line ending, but the same MD5 hash when we do? If so, then imposing a common line ending  seems worth trying as well, though as you say we'd probably want to test on a few different machines/operating systems to make sure we don't see unintended effects.
* On the fact that git(-lfs) perceives differences in binaries when they are produced on different machines (even though the underlying "content" is the same), my guess is that will be difficult to avoid and so probably not worth investigating further.

Let me know what you think, thanks.

**Comment by veli-m-andirin** [comment]
Thanks @jmshapir!

- Sounds great on imposing a stable sort rule.
- On imposing a common line ending, [here](https://github.com/JMSLab/Skills/issues/200#issuecomment-917416293) is the discussion where we confirmed that fixing line endings indeed helps. I will make some research about whether imposing a particular line ending can cause any problems. 
- I agree with your opinion on this.

I propose the following to-dos here:

- [ ] @veli-m-andirin makes code changes to implement a stable sorting rule.
- [ ] @veli-m-andirin makes some research on whether adopting a non-native line ending can cause problems on different platforms, and if not, implements this suggestion.

@santiagohermo @miikapaal @jmshapir please let me know your thoughts!

**Comment by jmshapir** [comment]
@veli-m-andirin the [plan you propose](https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1117542734) sounds good to me, thanks!

**Comment by veli-m-andirin** [comment]
@jmshapir @santiagohermo @miikapaal, hopefully one of the final updates here :)

I implemented stable sorting when sortbykey is TRUE in 203e930.

As for imposing a single line ending in the output files, I found [this article](https://www.aleksandrhovhannisyan.com/blog/crlf-vs-lf-normalizing-line-endings-in-git/) to be an excellent read. In short, when collaborating with Git, my sense is that people deal with line endings by letting Git to do the conversions through the `.gitattributes` file. In the recent past, Linux line endings caused problems even in common Windows software such as Notepad. In contrast, some `.gitattributes` settings such as `* text=auto` allow people to have native line endings locally while maintaning a single line ending remotely, which I think would prevent Git flagging changes due to line endings. 

I think it might not be necessary to start modifying `.gitattributes` in every single project right now, but we can keep this comment as a reference and if we ever come across a similar issue we can consider modifying `.gitattributes` in the relevant project. As for modifying SaveData, I think forcing a Linux line ending on a Windows machine or vice-versa don't sounds like great ideas. 

These being said, I think we completed the to-dos in https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1117542734 and can move to a pull for the minor change we made to the sorting here. Please let me know if you disagree.

**Comment by jmshapir** [comment]
Thanks @veli-m-andirin!

I'm on board with moving to a pull. Maybe @santiagohermo @miikapaal could serve as reviewers as they are familiar with the context here?

**Comment by jmshapir** [comment]
Thanks @santiagohermo!

@veli-m-andirin since @miikapaal gave the 👍 too I think we are good to move to a pull here!

**Comment by veli-m-andirin** [comment]
Summary: In this issue we investigated MD5 changes in files outputted by SaveData. As instabilities in ordering of rows based on OS differences seemed to be one of the reasons, we implemented a stable sorting. We also started to coerce objects into data tables before saving.

Changes were merged into master in 6071de9. Final state of the issue branch is [here](https://github.com/gslab-econ/gslab_r/tree/edc3ca538a59cf0a3b758c931d94f7ed83f39de6). Final state of the issue sub is [here](https://github.com/gslab-econ/gslab_r/tree/edc3ca538a59cf0a3b758c931d94f7ed83f39de6/issue15).
